### PR TITLE
Feature: gain scheduling

### PR DESCRIPTION
**Pull request by Deakensbubbles**
Closes PM-

## Description

A primitive form of gainscheduling able to change the P, I and D values of the controller based on the selected gait. at the moment the corresponding gain values to each gait are arbitrarily chosen. The " lookup table" used to chose these values is hardcoded in the node and will be replaced with a config file asap.

## Changes

A ROS package containing a Python package with a ROS Node that changes the PID values based on the selected gait.

### CODE REVIEW

File: march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    def gait_selection_callback(self, data):
+        rospy.logdebug("This is the gait name: %s", data.goal.name)
+        if self.gait_name != data.goal.name:
+            rospy.logdebug("The selected gait: {0} is not the same as the previous gait: {1}".format(
+                data.goal.name, self.gait_name))
+            self.gait_name = data.goal.name
+            self.client_update()
+
+    def listener(self):
+        rospy.Subscriber("march/gait/perform/goal", GaitNameActionGoal, callback=self.gait_selection_callback)
```
Comment (Olavhaasie) [inline march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py]: Why is this not defined in the `__init__`? Then you also do not have to call this method

File: march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import rospy
+from dynamic_reconfigure.client import Client
+from march_shared_resources.msg import GaitNameActionGoal
+
+
+class DynamicPIDReconfigurer:
+    def __init__(self, gait_name=None, joint_list=None):
+        self.gait_name = gait_name
+        self.joint_list = joint_list
+        self.my_clients = []
```
[the remark this answers is not in the document]
Comment (Deakensbubbles) [reply march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py]: I will change the names of these variables accordingly

File: march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py (hunk 1 above)
Comment (Deakensbubbles) [reply march_gain_scheduling/src/gain_scheduling/dynamic_pid_reconfigurer.py]: I find it more readable if the functionality of the main loop is apparent  due to functions being called. Not having any called functions of the class can make it seem like it is not being used except for creating an object.

File: march_gain_scheduling/src/march_gain_scheduling/dynamic_pid_reconfigurer.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def gait_selection_callback(self, data):
+        rospy.logdebug("This is the gait name: %s", data.goal.name)
+        if self._gait_name != data.goal.name:
+            rospy.logdebug("The selected gait: {0} is not the same as the previous gait: {1}".format(
+                data.goal.name, self._gait_name))
+            self._gait_name = data.goal.name
+            self.client_update()
+
+    def listener(self):
```
Comment (Roelemans) [inline march_gain_scheduling/src/march_gain_scheduling/dynamic_pid_reconfigurer.py]: I agree with Olav. Move to __init__. Get used to it, this is how we do things in the MARCH cartel. New qute noobies will just have to learn the proper format.

**Comment by Olavhaasie** [comment]
Something odd is happening with the Melodic build, I'll try to fix it :rescue_worker_helmet:

**Comment by Deakensbubbles** [comment]
Thank you Olav! I've been pulling my hairs out for a good hour now trying to find out how to fix this but, to no avail.

**Comment by Olavhaasie** [comment]
See #383 for a fix

**Pull request merged**